File: leetcode/20. 有效的括号.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let len = s.length;
    if (!len) return true;
    if (len % 2) return false;
    let stack = [];
    let map = {
        ")": "(",
        "]": "[",
        "}": "{",
    };
    for (let ch of s) {
        if (ch === ")" || ch === "]" || ch === "}") {
            if (stack.length && stack.pop() !== map[ch]) {
                return false;
            }
        } else {
            stack.push(ch);
        }
    }
    return !stack.length;
};

File: archive/leetcode/20. 有效的括号.cpp
/*
 * @Author       : ApassEr
 * @Date         : 2020-07-03 22:39:03
 * @LastEditTime : 2020-07-03 22:40:20
 * @Description  : 20. 有效的括号
 * @Algorithm    : 栈
 */
class Solution {
public:
    bool isValid(string s) {
        if (s.empty()) return true;
        stack<char> sc;
        unordered_map<char, char> m = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        for (auto i : s) {
            if (i == ')' || i == ']' || i == '}') {
                if (!sc.empty() && sc.top() == m[i]) {
                    sc.pop();
                } else {
                    return false;
                }
            } else {
                sc.push(i);
            }
        }
        return sc.empty() ? true : false;
    }
};